File: scripts/merge_sort.py
def merge(arr,l,m,r):
    n1 = (m-l)  + 1
    n2 = r-m

    L = [0]*(n1)
    R = [0]*(n2)

    for i in range(0,n1):
        L[i] = arr[l + i]

    for j in range(0, n2):
        R[j] = arr[m + 1 +j]

    i = 0
    j = 0
    k = l

    while(i < n1 and j < n2):
        if(L[i]<=R[j]):
            arr[k] = L[i]
            i = i+1
        else:
            arr[k] = R[j]
            j = j+1
        k = k+1

    while(i<n1):
        arr[k] = L[i]
        i = i+1
        k = k+1

    while(j<n2):
        arr[k] = R[j]
        j = j+1
        k = k+1

def mergeSort(arr,l,r):
    if(l<r):
        m = int((l+(r-1))/2)

        mergeSort(arr,l,m)
        mergeSort(arr,m+1,r)
        merge(arr,l,m,r)


arr = []
size = int(input("Enter the size"))

for i in range(0,size):
    a = int(input())
    arr.append(a)

n = len(arr)
print("List before sorting")
print(arr)

mergeSort(arr,0,n-1)
print("sorted list ")
print(arr)

//Made by Prakhar Varshney

File: mergesort.cpp
// Merge Sort
// this program generates a random array and sorts it using merge sort
// it also shows the ratio of (actual number of comparision) / (expected number of comparision)
#include <bits/stdc++.h>
using namespace std;
int c1 = 0, c2 = 0;
void generate(int *a, int size, int range)
{
    for (int i = 0; i < size; i++)
    {
        a[i] = (rand() % range);
    }
}

void merge(int a[], int p, int q, int r)
{

    int n1 = q - p + 1;
    int n2 = r - q;

    int L[n1], M[n2];

    for (int i = 0; i < n1; i++)
    {
        L[i] = a[p + i];
        // c1++;
    }
    for (int j = 0; j < n2; j++)
    {
        M[j] = a[q + 1 + j];
        // c1++;
    }
    int i, j, k;
    i = 0;
    j = 0;
    k = p;
    while (i < n1 && j < n2)
    {
        if (L[i] <= M[j])
        {
            a[k] = L[i];
            i++;
            c1++;
        }
        else
        {
            a[k] = M[j];
            j++;
            c1++;
        }
        k++;
    }
    while (i < n1)
    {
        a[k] = L[i];
        i++;
        k++;
    }

    while (j < n2)
    {
        a[k] = M[j];
        j++;
        k++;
    }
}

void mergeSort(int a[], int l, int r)
{
    if (l < r)
    {
        int q = l + (r - l) / 2;
        mergeSort(a, l, q);
        mergeSort(a, q + 1, r);
        merge(a, l, q, r);
    }
}

void printArray(int a[], int size)
{
    for (int i = 0; i < size; i++)
        cout << a[i] << " ";
    cout << endl;
}

int main()
{
    int a[] = {6, 5, 12, 10, 9, 1};
    // int size = sizeof(arr) / sizeof(arr[0]);
    int size, range;
    cout << "Enter size and range of array: \n";
    cin >> size >> range;

    generate(a, size, range);
    cout << "Before Sorting: \n";
    printArray(a, size);

    mergeSort(a, 0, size - 1);

    cout << "Sorted array: \n";
    printArray(a, size);

    c2 = size * log(size);
    cout << "c1: " << c1 << endl;
    cout << "c2: " << c2 << endl;
    cout << "c1/c2: " << (c1 * 1.0) / c2 << endl;
    return 0;
}
